cmd/containerd-shim-runhcs-v1/task_hcs.go

```diff
@@ -853,3 +840,129 @@ func (ht *hcsTask) Stats(ctx context.Context) (*stats.Statistics, error) {
+
+	if err := verifyTaskUpdateResourcesType(resources); err != nil {
+		return err
+	}
+
+	if ht.ownsHost && ht.host != nil {
+		return ht.host.UpdateConstraints(ctx, resources, req.Annotations)
+	}
+
+	return ht.updateTaskContainerResources(ctx, resources, req.Annotations)
+}
+
+func (ht *hcsTask) updateTaskContainerResources(ctx context.Context, data interface{}, annotations map[string]string) error {
+	if ht.isWCOW {
+		return ht.updateWCOWResources(ctx, data, annotations)
+	}
+
+	return ht.updateLCOWResources(ctx, data, annotations)
+}
+
+func (ht *hcsTask) updateWCOWContainerCPU(ctx context.Context, cpu *specs.WindowsCPUResources) error {
+	// if host is 20h2+ then we can make a request directly to hcs
+	if osversion.Get().Build >= osversion.V20H2 {
+		req := &hcsschema.Processor{}
+		if cpu.Count != nil {
+			req.Count = int32(*cpu.Count)
+		}
+		if cpu.Maximum != nil {
+			req.Maximum = int32(*cpu.Maximum)
+		}
+		if cpu.Shares != nil {
+			req.Weight = int32(*cpu.Shares)
+		}
+		return ht.requestUpdateContainer(ctx, resourcepaths.SiloProcessorResourcePath, req)
+	}
+
+	// host is < 20H2, so we can use our job object utility tool to get the same result
+	hostPath, guestPath := getToolPaths()
+	shareReq := &shimdiag.ShareRequest{
+		HostPath: hostPath,
+		UvmPath:  guestPath,
+		ReadOnly: true,
+	}
+	// Add utility tool to container host if not present
+	err := ht.Share(ctx, shareReq)
+	if err == errTaskNotIsolated {
+		// process isolated task, just use host path to tool
+		guestPath = hostPath
+	} else if err != nil {
+		return err
+	}
+
+	args := createJobObjectsUtilArgs(ctx, guestPath, cpu)
+	req := &shimdiag.ExecProcessRequest{
+		Args: args,
+	}
+	if exitCode, err := ht.ExecInHost(ctx, req); err != nil {
+		return errors.Wrapf(err, "failed to exec command %v in host: %d", args, exitCode)
+	}
+
+	return nil
+}
+
+func isValidWindowsCPUResources(c *specs.WindowsCPUResources) bool {
+	return (c.Count != nil && (c.Shares == nil && c.Maximum == nil)) ||
+		(c.Shares != nil && (c.Count == nil && c.Maximum == nil)) ||
+		(c.Maximum != nil && (c.Count == nil && c.Shares == nil))
+}
+
+func (ht *hcsTask) updateWCOWResources(ctx context.Context, data interface{}, annotations map[string]string) error {
+	resources, ok := data.(*specs.WindowsResources)
+	if !ok {
+		return errors.New("must have resources be type *WindowsResources when updating a wcow container")
+	}
+	if resources.Memory != nil && resources.Memory.Limit != nil {
+		newMemorySizeInMB := *resources.Memory.Limit / bytesPerMB
+		memoryLimit := normalizeContainerMemorySize(ctx, ht.id, newMemorySizeInMB)
+		if err := ht.requestUpdateContainer(ctx, resourcepaths.SiloMemoryResourcePath, memoryLimit); err != nil {
+			return err
+		}
+	}
+	if resources.CPU != nil {
+		if !isValidWindowsCPUResources(resources.CPU) {
+			return fmt.Errorf("invalid cpu resources request for container %s: %v", ht.id, resources.CPU)
+		}
+		if err := ht.updateWCOWContainerCPU(ctx, resources.CPU); err != nil {
```

Comment:
Just to make sure I'm understanding, before 20H2 there wasn't a way to update cpu limits directly via HCS so you call this to either 
1. Call into HCS if on a supported build
2. Use your job object tool instead

Updating memory was always supported I take it, just by the direct call always

Reply:
correct

Reply:
Thanks!